### ISSUE TITLE
Inconsistent tokenization with examples/server tokenize endpoint

### ISSUE THREAD

**Issue by IsaacDynamo** [ggerganov/llama.cpp#3287]
When using the `tokenize` endpoint of the `example/server` with `llama-2-7b-chat.Q5_K_M.gguf`, tokenization is inconsistent with the documentation.

*"Note that the special BOS token is not added in front of the text and also a space character is not inserted automatically as it is for /completion."* [doc](https://github.com/ggerganov/llama.cpp/tree/a5661d7e71d15b8dfc81bc0510ba912ebe85dfa3/examples/server#api-endpoints)

However it seems that the space is being added to the content.

The following show an incorrect round trip where `"Hello:World"` becomes `" Hello:World"`
```
$ curl -s --request POST --url http://localhost:8080/tokenize --header "Content-Type: application/json" 
--data '{"content": "Hello:World" }'
{"tokens":[15043,29901,14058]}
$ curl -s --request POST --url http://localhost:8080/detokenize --header "Content-Type: application/json" 
--data '{"tokens": [15043,29901,14058] }'
{"content":" Hello:World"}
```

The expected tokenization would have been `[10994, 29901, 14058]`, because that gives back `"Hello:World"`.
```
$ curl -s --request POST --url http://localhost:8080/detokenize --header "Content-Type: application/json" 
--data '{"tokens": [10994, 29901, 14058] }'
{"content":"Hello:World"}
```

The array variant of the `tokenize` endpoint even shows that a space is inserted for each substring. 
```
$ curl -s --request POST --url http://localhost:8080/tokenize --header "Content-Type: application/json" 
--data '{"content": ["Hello",":","World"] }'
{"tokens":[15043,584,2787]}
$ curl -s --request POST --url http://localhost:8080/detokenize --header "Content-Type: application/json" 
--data '{"tokens": [15043,584,2787] }'
{"content":" Hello : World"}
```

**TLDR**
I expected the `tokenize` and `detokenize` endpoints to be each others inverse. But that is currently not the case due to the automatic space insertion.

**Comment by staviq** [comment]
Server seems to be a bit behind recent changes, there are couple more problems than this, and it became slightly divergent from how `main` works.

I haven't tried it, but I've been thinking, maybe server could be modified to work more like `console.cpp` in `main`, and only deal with http side of things, letting `main` code do the work.

**Comment by IsaacDynamo** [comment]
After looking some more I found that the original paper only specifies `decode(encode(normalize(x))) == normalize(x)`.

However most implementations don't make a distinction between `encode()` and `normalize()`. Even the paper says that `normalize()` is done in `encode()`.
> "Encoder internally executes Normalizer to normalize the input text"

Furthermore some decoders also perform the "denormalize" step (`.replace('_', ' ')`), which breaks the above equality.

This is because most people want `denormalize(decode(encode(normalize(x)))) ~= x` which is reasonable.  So in most implementations the normalization and "denormalization" behavior has been embedded in `encode()` and `decode()`.  Such that `decode(encode(x)) ~= x`. 

Sometimes the (de)normalization details can be controlled with configuration flags to these routines.

So it might be useful to add an additional configuration flag that gives the following properties:
- `decode(encode(a + b)) == decode(encode(a)) + decode(encode(b))`
- `decode(encode(a + b)) == decode(encode(a) + encode(b))`

I think this can be done with a flag that disables the insertion of the space prefix.

Something similar is proposed in https://github.com/google/sentencepiece/pull/908

However adding such an additional flag to `llama_tokenize()` will be a breaking change. Is such a change acceptable?

**Comment by goerch** [comment]
> After looking some more I found that the original paper only specifies `decode(encode(normalize(x))) == normalize(x)`.

Having done some character level testing of the tokenizers I'm interested in testing these kind of invariants. Do you have a reference of the mentioned paper for us?  Thanks!

**Comment by IsaacDynamo** [comment]
The paper: [SentencePiece: A simple and language independent subword tokenizer
and detokenizer for Neural Text Processing](https://arxiv.org/pdf/1808.06226.pdf)

To be clear the paper only mentions `decode(encode(normalize(x))) == normalize(x)`. The other properties would be nice for composing prompts.
